### PR TITLE
GF coldpool updates and bugfix for radiation coupling

### PR DESCRIPTION
**Pull request by haiqinli**
Enter a description of this PR. This should include why this PR was created, and what it does.
-- A bugfix to add GF clouds on top of the clouds provided by the other options. The GF coldpool parameterization impacts on momentum is also updated. 

Testing and relations to other Pull Requests should be added as subsequent comments.
-- This PR has been tested on Jet.

### CODE REVIEW

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F [hunk 1]
```diff
@@ -38,7 +38,9 @@ MODULE module_cu_gf_deep
 !> still 16 ensembles for clousres
      integer, parameter:: maxens3=16
 !>  which coldpool version to use
-     integer, parameter:: icoldpool=1
+     integer, parameter:: icoldpool=1  ! 1 = momentum impact; 2 = sf flx impact
+                                       ! 3 = both, (1) and (2)
```
Comment (barlage) [inline src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F:43]: clearer to put 2 on its own line

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F [hunk 2]
```diff
@@ -2304,23 +2306,44 @@ subroutine cu_gf_deep_run(        &
           endif
       enddo
       !-----downdraft_cold_pool---------------
-      if (icoldpool > 0 .and. imid ==0) then
+      !if (icoldpool > 0 .and. imid ==0) then
```
Comment (barlage) [inline src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F]: remove commented line

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F [hunk 3]
```diff
@@ -2304,23 +2306,44 @@ subroutine cu_gf_deep_run(        &
...
       ! --- only magnitude is augmented, direction is kept the same
       do i=its,itf
+          umcl(i)=0.
+          vmcl(i)=0.
+          dpsum=0.
           if(ierr(i) > 0 ) cycle
+           do k=kts+1,ktop(i)-1
+             trash =-(po_cup(i,k)-po_cup(i,kts))
+             if(trash.gt.300. .and. trash.lt.600.)then
+             dp=100.*(po_cup(i,k)-po_cup(i,k+1))
```
Comment (barlage) [inline src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F]: indent for this if block

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F [hunk 4]
```diff
@@ -2304,23 +2306,44 @@ subroutine cu_gf_deep_run(        &
...
+             umcl(i)=umcl(i)+us(i,k)*dp
+             vmcl(i)=vmcl(i)+us(i,k)*dp
+             dpsum=dpsum+dp
+             endif
+           enddo
+           if(dpsum > 0.) then
+           umcl(i)=umcl(i)/dpsum
```
Comment (barlage) [inline src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F]: indent for this if block

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F [hunk 5]
```diff
@@ -2304,23 +2306,44 @@ subroutine cu_gf_deep_run(        &
...
-          if(ierr(i) > 0 ) cycle
-           do k=kts,kdet(i)
+          if(ierr(i) > 0 .or. vcpool(i) .le.0.) cycle
+             k=kts
              dp=100.*(po_cup(i,k)-po_cup(i,k+1))
-             outu(i,k) = outu(i,k) + edto(i)*zdo(i,k+1)*(umcl(i)-us(i,k))*g/dp*xmb(i)
-             outv(i,k) = outv(i,k) + edto(i)*zdo(i,k+1)*(vmcl(i)-vs(i,k))*g/dp*xmb(i)
+             outu(i,k) = outu(i,k) + edto(i)*zdo(i,k+1)*umcl(i)*g/dp*xmb(i)
+             outv(i,k) = outv(i,k) + edto(i)*zdo(i,k+1)*vmcl(i)*g/dp*xmb(i)
+                do k=kts+1,kdet(i)
```
Comment (barlage) [inline src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F]: indenting is messed up here

File: src/core_atmosphere/physics/mpas_atmphys_driver_cloudiness.F [hunk 6]
```diff
@@ -130,7 +130,6 @@ subroutine deallocate_cloudiness(configs)
     if(allocated(qccu_p)) deallocate(qccu_p)
  endif
 
-
```
Comment (barlage) [inline src/core_atmosphere/physics/mpas_atmphys_driver_cloudiness.F]: for future merge purposes, I would avoid removing lines in the mpas core code

File: src/core_atmosphere/physics/mpas_atmphys_driver_cloudiness.F [hunk 7]
```diff
@@ -232,25 +231,24 @@ subroutine cloudiness_from_MPAS(itimestep,configs,mesh,diag_physics,sfc_input,it
...
  
+ if (convection_scheme=="cu_grell_freitas") then
+    call mpas_pool_get_array(diag_physics,'cldfrac_cu',cldfrac_cu)
+    call mpas_pool_get_array(diag_physics,'qi_cu',qi_cu)
+    call mpas_pool_get_array(diag_physics,'qc_cu',qc_cu)
+    do j = jts,jte
+    do k = kts,kte
+    do i = its,ite
+       if (cldfrac_cu(k,i) .gt. 0.001) then
+          cldfrac_p(i,k,j) = min(1._RKIND,cldfrac_p(i,k,j) + cldfrac_cu(k,i))
```
Comment (barlage) [inline src/core_atmosphere/physics/mpas_atmphys_driver_cloudiness.F:244]: is the intention to always overwrite the mynn cldfrac calculated above when cu_grell_freitas?

File: src/core_atmosphere/physics/mpas_atmphys_driver_cloudiness.F [hunk 8]
```diff
@@ -347,16 +345,16 @@ subroutine driver_cloudiness(itimestep,configs,mesh,diag_physics,sfc_input,its,i
...
-       call mpas_timer_start('cal_gfcld')
-       call cal_gfcld(cldfrac_p , qcrad_p   , qirad_p    , &
-                      qccu_p    , qicu_p    ,              &
-                      its,  ite , jts , jte , kts , kte  )
-      call mpas_timer_stop('cal_gfcld')
-    endif
-
  end select cld_fraction_select
 
+ if (convection_scheme=="cu_grell_freitas") then
```
Comment (barlage) [inline src/core_atmosphere/physics/mpas_atmphys_driver_cloudiness.F:351]: similar to above, is the intention to always overwrite the fields calculated in the case statement above?

**Review by barlage — COMMENTED** [review]
A few formatting suggestions (take em or leave em). And a question about the intention of taking the two blocks out of the case statement

**Comment by barlage** [comment]
@haiqinli maybe I understand now that these are additive on top of the other schemes.

File: src/core_atmosphere/physics/mpas_atmphys_driver_cloudiness.F (hunk 8 above)
Comment (barlage) [reply src/core_atmosphere/physics/mpas_atmphys_driver_cloudiness.F:351]: I looked at cal_gfcld and understand now what is happening.

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F (hunk 3 above)
Comment (haiqinli) [reply src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F]: The if block is used to only apply the convective clouds between 300hPa and 600hPa above surface.

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F (hunk 4 above)
Comment (haiqinli) [reply src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F]: The sum of cloud depth (dpsum) should be large then 0 (dpsum is initialized as 0).

**Comment by haiqinli** [comment]
@barlage Yes, the convective clouds are additive on top of the other schemes. Joe pointed out this issue yesterday. I will also update the formatting following your suggestions. Thanks!

**Review by joeolson42 — APPROVED** [review]
Approved after the minor formatting changes called out by Mike.

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F (hunk 3 above)
Comment (barlage) [reply src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F]: my suggestion is to indent these lines in the if block

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F (hunk 4 above)
Comment (barlage) [reply src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F]: same, just a suggestion to indent the lines in the if block

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F (hunk 1 above)
Comment (haiqinli) [reply src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F:43]: Updated following your suggestions. Thanks.

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F (hunk 2 above)
Comment (haiqinli) [reply src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F]: Removed following your suggestions. Thanks.

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F (hunk 3 above)
Comment (haiqinli) [reply src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F]: Updated following your suggestions. Thanks.

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F (hunk 4 above)
Comment (haiqinli) [reply src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F]: Updated following your suggestions. Thanks.

File: src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F (hunk 5 above)
Comment (haiqinli) [reply src/core_atmosphere/physics/physics_wrf/module_cu_gf_deep.F]: Updated following your suggestions. Thanks.

**Comment by jderrico-noaa** [comment]
@barlage do the changes sufficiently address the formatting concerns or should I hold off on merging?

**Review by tanyasmirnova — APPROVED** [review]

**Review by barlage — APPROVED** [review]

**Pull request merged**